src/coreclr/jit/lowerarmarch.cpp

```diff
@@ -2266,156 +2270,105 @@ void Lowering::ContainCheckCompare(GenTreeOp* cmp)
 
 #ifdef TARGET_ARM64
 //------------------------------------------------------------------------
-// IsValidCompareChain : Determine if the node contains a valid chain of ANDs and CMPs.
+// TryLowerAndToCCMP : Lower an and of two conditions into test + CCMP + SETCC nodes.
 //
 // Arguments:
-//    child - pointer to the node being checked.
-//    parent - parent node of the child.
-//
-// Return value:
-//    True if a valid chain is found.
+//    node - pointer to the node
 //
-// Notes:
-//    A compare chain is a sequence of CMP nodes connected by AND nodes.
-//    For example:   AND (AND (CMP A B) (CMP C D)) (CMP E F)
-//    The chain can just be a single compare node, however it's parent
-//    must always be an AND or SELECT node.
-//    If a CMP or AND node is contained then it and all it's children are
-//    considered to be in a valid chain.
-//    Chains are built up during the lowering of each successive parent.
-//
-bool Lowering::IsValidCompareChain(GenTree* child, GenTree* parent)
+GenTree* Lowering::TryLowerAndToCCMP(GenTree* tree)
 {
-    assert(parent->OperIs(GT_AND) || parent->OperIs(GT_SELECT));
+    assert(tree->OperIs(GT_AND));
 
-    if (parent->isContainedCompareChainSegment(child))
+    if (!comp->opts.OptimizationEnabled())
     {
-        // Already have a chain.
-        return true;
+        return nullptr;
     }
-    else if (child->OperIs(GT_AND))
+
+    GenTree* op1 = tree->gtGetOp1();
+    GenTree* op2 = tree->gtGetOp2();
+
+    if (!op1->OperIsCompare() && !op1->OperIs(GT_SETCC))
     {
-        // Count both sides.
-        return IsValidCompareChain(child->AsOp()->gtGetOp2(), child) &&
-               IsValidCompareChain(child->AsOp()->gtGetOp1(), child);
+        return nullptr;
     }
-    else if (child->OperIsCmpCompare() && varTypeIsIntegral(child->gtGetOp1()) && varTypeIsIntegral(child->gtGetOp2()))
+
+    if (!op2->OperIsCmpCompare() || !varTypeIsIntegral(op2->gtGetOp1()))
```

Comment:
Wondering why this is also not `!op2->operIsCompare()`, just like you have `!op1->OperIsCompare()` above?

Reply:
The operands of `op2` will be compared with the `ccmp` instruction, and we can only do that for the "basic" relops: `GT_EQ`, `GT_NE`, `GT_LE`, `GT_LT`, `GT_GT`, `GT_GE`. The remaining compare operators, `GT_TEST_EQ` and `GT_TEST_NE` (whose semantics are `(a & b) == 0` and `(a & b) != 0`) cannot be implemented as a conditional compare -- it would need some kind of `ctst` instruction.

On the other hand, the first operand can be anything that has already set the condition flags in any way, including e.g. floating point compares or `tst` instruction.

Let me add a more detailed comment here about this

Reply:
Added a more detailed comment. I also experimented with making this logic symmetric (allow `ccmp` for either `op1` or `op2`), but diffs showed that it didn't matter. Might still do it as part of the `OR` follow up, though.